core/trino-main/src/main/java/io/trino/operator/join/BigintPagesHash.java

```diff
@@ -0,0 +1,198 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator.join;
+
+import com.google.common.collect.ImmutableList;
+import io.airlift.units.DataSize;
+import io.trino.operator.HashArraySizeSupplier;
+import io.trino.operator.PagesHashStrategy;
+import io.trino.spi.Page;
+import io.trino.spi.PageBuilder;
+import io.trino.spi.block.Block;
+import it.unimi.dsi.fastutil.longs.LongArrayList;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.Arrays;
+import java.util.List;
+
+import static io.airlift.slice.SizeOf.sizeOf;
+import static io.airlift.units.DataSize.Unit.KILOBYTE;
+import static io.trino.operator.SyntheticAddress.decodePosition;
+import static io.trino.operator.SyntheticAddress.decodeSliceIndex;
+import static io.trino.operator.join.PagesHash.getHashPosition;
+import static io.trino.util.HashCollisionsEstimator.estimateNumberOfHashCollisions;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+
+// This implementation assumes:
+// -There is only one join channel and it is of type bigint
+// -arrays used in the hash are always a power of 2.
+public final class BigintPagesHash
+        implements PagesHash
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(BigintPagesHash.class).instanceSize();
+    private static final DataSize CACHE_SIZE = DataSize.of(128, KILOBYTE);
+
+    private final LongArrayList addresses;
+    private final List<Block> joinChannelBlocks;
+    private final PagesHashStrategy pagesHashStrategy;
+
+    private final int mask;
+    private final int[] key;
+    private final long[] values;
+    private final long size;
+
+    private final long hashCollisions;
+    private final double expectedHashCollisions;
+
+    public BigintPagesHash(
+            LongArrayList addresses,
+            PagesHashStrategy pagesHashStrategy,
+            PositionLinks.FactoryBuilder positionLinks,
+            HashArraySizeSupplier hashArraySizeSupplier,
+            List<Page> pages,
+            int joinChannel)
+    {
+        this.addresses = requireNonNull(addresses, "addresses is null");
+        this.pagesHashStrategy = requireNonNull(pagesHashStrategy, "pagesHashStrategy is null");
+        requireNonNull(pages, "pages is null");
+        ImmutableList.Builder<Block> joinChannelBlocksBuilder = ImmutableList.builder();
+        for (Page page : pages) {
+            joinChannelBlocksBuilder.add(page.getBlock(joinChannel));
+        }
+        joinChannelBlocks = joinChannelBlocksBuilder.build();
+
+        // reserve memory for the arrays
+        int hashSize = hashArraySizeSupplier.getHashArraySize(addresses.size());
+
+        mask = hashSize - 1;
+        key = new int[hashSize];
+        values = new long[addresses.size()];
+        Arrays.fill(key, -1);
+
+        // We will process addresses in batches, to save memory on array of hashes.
```

Comment:
Ping about https://github.com/trinodb/trino/pull/13352#discussion_r945228353

Reply:
Sorry, I forgot about this. So the comment is basically not true. The batching here is done to improve spatial and temporal memory locality (and it actually does it significantly). This has nothing to do with the single-byte hashes.
I added this to my todo list for these classes

Reply:
I don't think that updating the comment in this class to be accurate should be a TODO as this is new code and the comment is completely untrue here while it is still partially true in the existing code.
Updating the comment in existing code can be follow-up.

Reply:
OK, I'll fix that after the CI finishes